elfo-core/src/message/any.rs

```diff
@@ -0,0 +1,613 @@
+            .ok_or(de::Error::invalid_length(1usize, &"tuple of 3 elements"))?;
+
+        de::SeqAccess::next_element_seed(&mut seq, MessageTag { protocol, name })?
+            .ok_or(de::Error::invalid_length(2usize, &"tuple of 3 elements"))
+    }
+}
+
+struct MessageTag<'a> {
+    protocol: &'a str,
+    name: &'a str,
+}
+
+impl<'de, 'tag> de::DeserializeSeed<'de> for MessageTag<'tag> {
+    type Value = AnyMessage;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let Self { protocol, name } = self;
+
+        let vtable = MessageVTable::lookup(protocol, name)
+            .ok_or_else(|| de::Error::custom(format_args!("unknown message: {protocol}/{name}")))?;
+
+        let out_ptr = alloc_repr(vtable);
+
+        let mut deserializer = <dyn erased_serde::Deserializer<'_>>::erase(deserializer);
+        // SAFETY: `out_ptr` belongs to the same object as the vtable.
+        unsafe { (vtable.deserialize_any)(&mut deserializer, out_ptr) }
+            .map_err(de::Error::custom)?;
+
+        Ok(AnyMessage(out_ptr))
+    }
+}
+
+cfg_network!({
+    use rmp_serde::{decode, encode};
+
+    impl AnyMessage {
+        #[doc(hidden)]
+        #[inline]
+        pub fn read_msgpack(
+            buffer: &[u8],
+            protocol: &str,
+            name: &str,
+        ) -> Result<Option<Self>, decode::Error> {
+            let Some(vtable) = MessageVTable::lookup(protocol, name) else {
+                return Ok(None);
+            };
+
+            let out_ptr = alloc_repr(vtable);
+
+            // SAFETY: `out_ptr` belongs to the same object as the vtable.
+            unsafe { (vtable.read_msgpack)(buffer, out_ptr) }?;
+
+            Ok(Some(Self(out_ptr)))
+        }
+
+        #[doc(hidden)]
+        #[inline]
+        pub fn write_msgpack(&self, out: &mut Vec<u8>, limit: usize) -> Result<(), encode::Error> {
+            let vtable = self._vtable();
+            // SAFETY: the vtable belongs to `self`.
+            unsafe { (vtable.write_msgpack)(self.0, out, limit) }
+        }
+    }
+});
+
+// === AnyMessageRef ===
+
+/// A reference to the message inside [`Envelope`] or [`AnyMessage`].
+///
+/// [`Envelope`]: crate::Envelope
+pub struct AnyMessageRef<'a> {
+    inner: ManuallyDrop<AnyMessage>, // never drop, borrows memory
+    marker: PhantomData<&'a AnyMessage>,
+}
+
+impl<'a> AnyMessageRef<'a> {
+    /// # Safety
+    ///
+    /// `ptr` must be a valid pointer for reads.
+    pub(crate) unsafe fn new(ptr: NonNull<MessageRepr>) -> Self {
+        Self {
+            inner: ManuallyDrop::new(AnyMessage(ptr)),
+            marker: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub fn downcast_ref<M: Message>(&self) -> Option<&'a M> {
+        let ret = self.inner.downcast_ref();
+
+        // SAFETY: we produce lifetime bound to the original one.
+        // Note: semantically it's shortening not extending.
```

Comment:
I’m not sure I get why it’s shortening. It’s transmuting `'self` to `'a`, where `&'self &'a` reference exists, so `'self` can’t be shorter than `'a`.

That being said, I think this operation is sound as long as `AnyMessage` is pinned in memory, which it seems to be.

Reply:
You're right, it's an incorrect comment